File: native/presentation/src/main/kotlin/com/wardrobapp/presentation/OutfitRating.kt
package com.wardrobapp.presentation

import java.util.Locale
import kotlin.math.min
import kotlin.math.roundToInt

/**
 * What a set of ratings adds up to.
 *
 * Ported from `src/domain/outfit-rating.ts` and held to it by
 * `outfit-rating.jsonl`. That module exists because the React Native app computed
 * this twice, differently: the outfit detail screen reduced the ratings it had
 * loaded and treated "none" as zero, while an unused service function asked
 * SQLite for `AVG(rating)` and treated "none" as null.
 */

/** The highest a rating goes, and so the most stars that can be filled. */
const val MAX_RATING = 5

data class RatingSummary(
    val count: Int,
    /** The mean, or null when there is nothing to average. */
    val average: Double?,
    /**
     * Stars to fill, 0 to [MAX_RATING].
     *
     * Rounded rather than truncated: an outfit rated 4 and 5 averages 4.5, and
     * showing four stars for that reads as the lower of the two opinions.
     */
    val stars: Int,
    /** The mean to one decimal place, or null when there is none. */
    val label: String?,
    /** Whether there is an average worth showing at all. */
    val showsAverage: Boolean,
)

fun ratingSummary(ratings: List<Int>): RatingSummary {
    // Ratings outside the scale cannot come from the star row, but they can come
    // from a restored backup or a hand-edited database, and an average of 9 would
    // fill more stars than exist. A zero means unrated rather than terrible.
    val usable = ratings.filter { it > 0 }

    if (usable.isEmpty()) {
        return RatingSummary(
            count = 0,
            average = null,
            stars = 0,
            label = null,
            showsAverage = false,
        )
    }

    val average = usable.sum().toDouble() / usable.size

    return RatingSummary(
        count = usable.size,
        average = average,
        stars = min(average, MAX_RATING.toDouble()).roundToInt(),
        label = oneDecimalPlace(average),
        showsAverage = true,
    )
}

/**
 * A number to one decimal place, the way JavaScript's `toFixed(1)` writes it.
 *
 * Pinned to [Locale.ROOT] rather than the device's: this is compared against what
 * the TypeScript produced, and half of Europe would render the separator as a
 * comma and fail a fixture for a reason that has nothing to do with the
 * arithmetic. The screen decides how to present it; this decides what it says.
 */
private fun oneDecimalPlace(value: Double): String =
    String.format(Locale.ROOT, "%.1f", value)

File: src/domain/outfit-rating.ts
/**
 * What a set of ratings adds up to.
 *
 * Pulled out because the app already computes it twice, differently. The outfit
 * detail screen reduces the ratings it loaded and treats "none" as zero, while
 * `getAverageRating` in the outfit service asks SQLite for `AVG(rating)` and
 * treats "none" as null — and nothing calls the second one. Two answers to one
 * question, neither tested.
 *
 * This is the answer. The rounding and the formatting live here too: the star row
 * shows a whole number of stars and the caption shows one decimal place, and both
 * are decisions about what the number means rather than about layout.
 */

export interface RatingSummary {
  /** How many ratings there are. */
  count: number;

  /** The mean, or null when there is nothing to average. */
  average: number | null;

  /**
   * Stars to fill, 0 to 5.
   *
   * Rounded rather than truncated: an outfit rated 4 and 5 averages 4.5, and
   * showing four stars for that reads as the lower of the two opinions.
   */
  stars: number;

  /** The mean to one decimal place, or null when there is none. */
  label: string | null;

  /** Whether there is an average worth showing at all. */
  showsAverage: boolean;
}

/** The highest a rating goes, and so the most stars that can be filled. */
export const MAX_RATING = 5;

export function ratingSummary(ratings: number[]): RatingSummary {
  // Ratings outside the scale cannot come from the star row, but they can come
  // from a restored backup or a hand-edited database, and an average of 9 would
  // fill more stars than exist.
  const usable = ratings.filter(rating => Number.isFinite(rating) && rating > 0);

  if (usable.length === 0) {
    return { count: 0, average: null, stars: 0, label: null, showsAverage: false };
  }

  const average = usable.reduce((sum, rating) => sum + rating, 0) / usable.length;
  const clamped = Math.min(average, MAX_RATING);

  return {
    count: usable.length,
    average,
    stars: Math.round(clamped),
    label: average.toFixed(1),
    showsAverage: true,
  };
}
